### PR TITLE
unix: loop starvation on successful write complete

### PR DESCRIPTION
**Pull request by jBarz**
A file descriptor that can do multiple successful write completion requests in a row will starve the loop because it will keep feeding the write_completed_queue. This fix will only process items on the write_completed_queue once per event loop (in uv__run_pending). Any new items on the queue will be processed in the next loop.

### CODE REVIEW

**Comment by gireeshpunathil** [comment]
will this cause performance degradation? At a time how many write requests (assuming infinite I/O readiness) will be committed with the new approach?I guess `uv__read` takes 32 iterations before it returns back to the loop to balance between read efficiency vs. loop availability.

**Comment by jBarz** [comment]
@gireeshpunathil Thats right. It will cause a performance degradation in the case where the next write request is invoked inside of the callback of the current write request.
```
do_write1 (){
  uv_write(write_req1, <small buffer>, write1_cb);
}
write1_cb(){ do_write2() }
do_write2 (){
  uv_write(write_req2, <small buffer>, ...);
}
```

`uv_write` will put the write_req2 on the `write_complete_queue` immediately if it succeeds.

Before this PR, write_req1 and write_req2 will happen consecutively from within `uv__write_callbacks`
After this PR, there will be a call to the platform polling function in between to check for other IO, timers etc.

**Comment by bnoordhuis** [comment]
In `uv__backend_poll()` we check a counter to avoid a similar issue.  You could combine that with this PR to prevent non-termination but without impacting benign cases too much.

**Comment by jBarz** [comment]
Do you mean `uv__io_poll`? Couldn't find uv__backend_poll.
So you're suggesting looping through the `write_complete_queue` for up to N times for every loop?
What should N be in that case?
I found `48` being used in `uv__io_poll`. Should we use the same here?

**Comment by bnoordhuis** [comment]
Sorry yes, `uv__io_poll()`.  I don't know if we have benchmarks that exercise this particular code path.  Maybe you could add one or alternatively check the running time of node's `make test` with your patch applied.

**Comment by jBarz** [comment]
Running on linux on s390x 
Without patch:
real    2m55.469s
user    4m9.760s
sys     0m39.087s

With patch:

real    2m52.046s
user    4m15.634s
sys     0m36.914s

**Comment by jBarz** [comment]
Resolved the z/OS error by rebasing on the latest v1.x
The freebsd error is with test/test-hrtime which doesn't seem related to this change.

**Comment by jBarz** [comment]
FYI, the benchmark `tcp_pump1_client` seems to run into this scenario.
It contains 1 client firing as much data as possible to 1 server with a timer to end the benchmark after approx 5 seconds.
The timer never gets called due to all writes being successful. So the benchmark times out after 60 seconds.
With my patch, it runs to completion after about 5-6 seconds.

**Comment by jBarz** [comment]
Before merging this, I am wondering if I should try implementing this TODO I found in `uv__write`
/* TODO: start trying to write the next request. */

This means doing the same thing as `uv__read`. If the IO operation is completely successful, try doing the next write but only upto 32 successive times.
Any opinions?

**Comment by jBarz** [comment]
I pushed one more change on top of the approved commit.
This change modifies `uv__write` to write to an infintely write-ready file descriptor up to 32 times (similar to `uv__read`). If I should create a separate PR for this, let me know.
Timing for `make test` on node master.

before 
real    2m21.658s
user    4m8.926s
sys     0m35.539s

after
real    2m21.943s
user    4m9.160s
sys     0m37.077s

**Comment by gireeshpunathil** [comment]
here: [link](https://gist.github.com/gireeshpunathil/a2fc2904675906f15efd3c93f7238e45)

**Comment by jBarz** [comment]
@gireeshpunathil I notice that the change to `uv__write` did not improve performance by much.
What I will do is revert the `uv__write` change and open another PR for that if neccessary.
I also included a test case for the original issue.
The test continously writes to a tcp server until a timer triggers in 2 seconds which shutsdown the connection.
Without the original fix, the test case times out.

**Comment by jBarz** [comment]
I also included a test case for the original issue.
The test continously writes to a tcp server until a timer triggers in 2 seconds which shutsdown the connection. Without the original fix, the test case times out.

**Comment by gireeshpunathil** [comment]
LGTM

File: test/test-tcp-write-ready.c [hunk 1]
```diff
@@ -0,0 +1,210 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Comment (cjihrig) [inline test/test-tcp-write-ready.c]: I think this should just be libuv collaborators, but double check a recently added file.

Comment (jBarz) [reply test/test-tcp-write-ready.c]: Thanks, Fixed

**Comment by gireeshpunathil** [comment]
I guess the main difference between `uv__read` and `uv__write` is that the former reads in 64KB chunks (it may be possible the kernel buffer has more than 64KB available, but as the data come from the remote endpoint it does not know how much to read in one shot), local iteration benefits. Whereas in the later, all the source data is attempted in one shot with the spill-over scheduled for the next loop iteration, so `uv__write` is inherently optimal. If `write` writes lesser than requested, it only means that  kernel buffer is full, and needs to wait till it drains to the wire.

So I guess we are good here.

File: test/test-tcp-write-ready.c [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+  uv_run(uv_default_loop(), UV_RUN_DEFAULT);
+
+  ASSERT(shutdown_cb_called == 1);
+  ASSERT(connect_cb_called == 1);
+  ASSERT(write_cb_called > 0);
+  ASSERT(close_cb_called == 1);
+
+  MAKE_VALGRIND_HAPPY();
+  return 0;
+}
```
Comment (bnoordhuis) [inline test/test-tcp-write-ready.c]: Can you merge this into `test/test-tcp-open.c`?  Duplicates too much code otherwise, IMO.

Comment (jBarz) [reply test/test-tcp-write-ready.c]: good idea. Will do

Comment (jBarz) [reply test/test-tcp-write-ready.c]: Done

File: test/test-tcp-open.c [hunk 3]
```diff
@@ -158,9 +201,34 @@ static void connect_cb(uv_connect_t* req, int status) {
...
+  r = uv_timer_start(&tm, timer_cb, 2000, 0);
+  ASSERT(r == 0);
+
+  r = uv_write(&write_req, stream, &buf, 1, write_cb);
+  ASSERT(r == 0);
+
+  /* Start reading */
+  r = uv_read_start(stream, alloc_cb, read_cb);
+  ASSERT(r == 0);
+}
```
Comment (bnoordhuis) [inline test/test-tcp-open.c]: Nothing seems to use these new functions?

Comment (jBarz) [reply test/test-tcp-open.c]: Sorry, fixed.

File: test/test-tcp-open.c [hunk 4]
```diff
@@ -122,6 +124,21 @@ static void read_cb(uv_stream_t* tcp, ssize_t nread, const uv_buf_t* buf) {
 }
 
 
+static void read1_cb(uv_stream_t* tcp, ssize_t nread, const uv_buf_t* buf) {
+  ASSERT(tcp != NULL);
+
+  if (nread >= 0) {
+    for (int i = 0; i < nread; ++i)
```
Comment (bnoordhuis) [inline test/test-tcp-open.c]: Can you use `for (i = ...; ...; ...)` here?  I'm reasonably sure `int i = ...` will be rejected by compilers in strict C89 mode.

Comment (jBarz) [reply test/test-tcp-open.c]: fixed

**Comment by bnoordhuis** [comment]
Looks like I wasn't wrong about compilers not liking it.  On most of the bots:

> test/test-tcp-open.c:131:5: error: 'for' loop initial declarations are only allowed in C99 mode

New CI: https://ci.nodejs.org/job/libuv-test-commit/810/

**Comment by jBarz** [comment]
Ping for CI :-)

**Comment by bnoordhuis** [comment]
Sorry, here you go: https://ci.nodejs.org/job/libuv-test-commit/821/

**Comment by jBarz** [comment]
Sorry I dropped the ball on this. One more CI and I should be able to respond immediately.

**Comment by cjihrig** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/881/

**Comment by bnoordhuis** [comment]
https://ci.nodejs.org/job/libuv-test-commit/886/

File: test/test-tcp-open.c [hunk 5]
```diff
@@ -122,6 +125,22 @@ static void read_cb(uv_stream_t* tcp, ssize_t nread, const uv_buf_t* buf) {
...
 
+static void read1_cb(uv_stream_t* tcp, ssize_t nread, const uv_buf_t* buf) {
+  int i;
+  ASSERT(tcp != NULL);
+
+  if (nread >= 0) {
+    for (i = 0; i < nread; ++i)
+      ASSERT(buf->base[i] == 'P');
+  }
+  else {
```
Comment (cjihrig) [inline test/test-tcp-open.c]: `else` on the previous line please.

File: test/test-tcp-open.c [hunk 6]
```diff
@@ -122,6 +125,22 @@ static void read_cb(uv_stream_t* tcp, ssize_t nread, const uv_buf_t* buf) {
...
+    ASSERT(nread == UV_EOF);
+    printf("GOT EOF\n");
```
Comment (cjihrig) [inline test/test-tcp-open.c]: Can the `printf()` be dropped?

File: test/test-tcp-open.c [hunk 7]
```diff
@@ -134,6 +153,36 @@ static void write_cb(uv_write_t* req, int status) {
 }
 
 
+static void write1_cb(uv_write_t* req, int status) {
+  uv_buf_t buf = uv_buf_init("P", 1);
```
Comment (cjihrig) [inline test/test-tcp-open.c]: Can you split the assignment from the declaration?

File: test/test-tcp-open.c [hunk 8]
```diff
@@ -134,6 +153,36 @@ static void write_cb(uv_write_t* req, int status) {
...
+  ASSERT(req != NULL);
```
Comment (cjihrig) [inline test/test-tcp-open.c]: Can you move this after the `int r;`

File: test/test-tcp-open.c [hunk 9]
```diff
@@ -158,9 +207,34 @@ static void connect_cb(uv_connect_t* req, int status) {
 }
 
 
+static void connect1_cb(uv_connect_t* req, int status) {
+  uv_buf_t buf = uv_buf_init("P", 1);
```
Comment (cjihrig) [inline test/test-tcp-open.c]: Can you separate the declaration and assignment here too.

**Comment by cjihrig** [comment]
Actually, it looks like AIX failures could be related.

**Comment by jBarz** [comment]
Addressed all nits.
There are 2 tests on AIX that failed. 
- tcp_write_fail
- tcp_write_ready (which is created by this PR)

In both tests, failure happened in the `uv_tcp_connect` callback.
Seems to be that the connection to the echo server is failing. Therefore I don't think it is related to the changes in this PR.
Also, I am unable to reproduce the failure on my own aix machine.
@mhdawson Could I get access to this aix machine and do further tests on it to see why the tcp connects are failing?

**Comment by jBarz** [comment]
I have run test tests multiple times on the very same machine and am unable to reproduce the failure.
I suspect that the failures in the last CI were a one-off.
Could we try another CI again?

**Comment by cjihrig** [comment]
New CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/909/

FWIW, AIX seems to be having a lot of problems lately, and the failures might not even be related to these changes.

**Comment by cjihrig** [comment]
The macOS timeout should be fixed by #1867. The only other failure was the same test on AIX. I think this should be good to go.

**Comment by jBarz** [comment]
rebased and squashed to single commit.
New CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/978/

**Comment by cjihrig** [comment]
Landed in https://github.com/libuv/libuv/commit/27e7a8b107538c65613d91324ecbbdaa1ecfed0b. Thanks!

**Pull request closed without merging**